src/common/datasource/src/file_format.rs

```diff
@@ -195,33 +206,125 @@ pub async fn infer_schemas(
 }
 
-pub async fn stream_to_file<T: DfRecordBatchEncoder, U: Fn(SharedBuffer) -> T>(
+/// Writes data to a compressed writer if the data is not empty.
+///
+/// Does nothing if `data` is empty; otherwise writes all data and returns any error.
+async fn write_to_compressed_writer(
+    compressed_writer: &mut CompressedWriter,
+    data: &[u8],
+) -> Result<()> {
+    if !data.is_empty() {
+        compressed_writer
+            .write_all(data)
+            .await
+            .context(error::AsyncWriteSnafu)?;
+    }
+    Ok(())
+}
+
+/// Streams [SendableRecordBatchStream] to a file with optional compression support.
+/// Data is buffered and flushed according to the given `threshold`.
+/// Ensures that writer resources are cleanly released and that an empty file is not
+/// created if no rows are written.
+///
+/// Returns the total number of rows successfully written.
+pub async fn stream_to_file<E>(
     mut stream: SendableRecordBatchStream,
     store: ObjectStore,
     path: &str,
     threshold: usize,
     concurrency: usize,
-    encoder_factory: U,
-) -> Result<usize> {
+    compression_type: CompressionType,
+    encoder_factory: impl Fn(SharedBuffer) -> E,
+) -> Result<usize>
+where
+    E: DfRecordBatchEncoder,
+{
+    // Create the file writer with OpenDAL's built-in buffering
+    let writer = store
+        .writer_with(path)
+        .concurrent(concurrency)
+        .chunk(DEFAULT_WRITE_BUFFER_SIZE.as_bytes() as usize)
+        .await
+        .with_context(|_| error::WriteObjectSnafu { path })?
+        .into_futures_async_write()
+        .compat_write();
+
+    // Apply compression if needed
+    let mut compressed_writer = writer.into_compressed_writer(compression_type);
+
+    // Create a buffer for the encoder
     let buffer = SharedBuffer::with_capacity(threshold);
-    let encoder = encoder_factory(buffer.clone());
-    let mut writer = LazyBufferedWriter::new(threshold, buffer, encoder, path, |path| async {
-        store
-            .writer_with(&path)
-            .concurrent(concurrency)
-            .chunk(DEFAULT_WRITE_BUFFER_SIZE.as_bytes() as usize)
-            .await
-            .map(|v| v.into_futures_async_write().compat_write())
-            .context(error::WriteObjectSnafu { path })
-    });
+    let mut encoder = encoder_factory(buffer.clone());
 
     let mut rows = 0;
 
+    // Process each record batch
     while let Some(batch) = stream.next().await {
         let batch = batch.context(error::ReadRecordBatchSnafu)?;
-        writer.write(&batch).await?;
+
+        // Write batch using the encoder
+        encoder.write(&batch)?;
         rows += batch.num_rows();
+
+        while buffer.buffer.lock().unwrap().len() >= threshold {
+            let chunk = {
+                let mut buffer_guard = buffer.buffer.lock().unwrap();
+                buffer_guard.split_to(threshold)
+            };
+            write_to_compressed_writer(&mut compressed_writer, &chunk).await?;
+        }
```

Comment:
It's better to lock it just once.
```suggestion
        loop {
            let chunk = {
                let mut buffer_guard = buffer.buffer.lock().unwrap();
                if buffer_guard.len() < threshold {
                    break;
                }
                buffer_guard.split_to(threshold)
            };
            write_to_compressed_writer(&mut compressed_writer, &chunk).await?;
        }
```

Reply:
You are right.
Fixed unreasonable locking logic.
The tests have been run locally and passed.